libkbfs/identify_util.go

```diff
@@ -6,12 +6,111 @@ package libkbfs
 
 import (
 	"fmt"
+	"sync"
 
 	"github.com/keybase/client/go/libkb"
 	"github.com/keybase/client/go/protocol/keybase1"
 	"golang.org/x/net/context"
+	"golang.org/x/sync/errgroup"
 )
 
+type extendedIdentify struct {
+	behavior   keybase1.TLFIdentifyBehavior
+	userBreaks chan keybase1.TLFUserBreak
+
+	tlfBreaksLock sync.Mutex
+	tlfBreaks     *keybase1.TLFBreak
+}
+
+func (ei *extendedIdentify) userBreak(username libkb.NormalizedUsername, uid keybase1.UID, breaks *keybase1.IdentifyTrackBreaks) {
+	if ei.userBreaks == nil {
+		return
+	}
+
+	ei.userBreaks <- keybase1.TLFUserBreak{
+		Breaks: breaks,
+		User: keybase1.User{
+			Uid:      uid,
+			Username: string(username),
+		},
+	}
+}
+
+func (ei *extendedIdentify) makeTlfBreaksIfNeeded(numUserInTlf int) {
+	if ei.userBreaks == nil {
+		return
+	}
+
+	b := &keybase1.TLFBreak{}
+	for i := 0; i < numUserInTlf; i++ {
+		ub := <-ei.userBreaks
+		if ub.Breaks != nil {
+			b.Breaks = append(b.Breaks, ub)
+		}
+	}
+
+	ei.tlfBreaksLock.Lock()
+	defer ei.tlfBreaksLock.Unlock()
+	ei.tlfBreaks = b
+}
+
+// getTlfBreakOrBust returns a keybase1.TLFBreak. This should only be called
+// for behavior.WarningInsteadOfErrorOnBrokenTracks() == true, and after
+// makeTlfBreaksIfNeeded is called. Otherwise it panics.
+func (ei *extendedIdentify) getTlfBreakOrBust() keybase1.TLFBreak {
```

Comment:
Maybe this should also `close(ei.userBreak)` to make sure no one else calls `makeTlfBreaksIfNeeded` after the response has been returned?  Or maybe have a separate func for it?  Just a minor cleanup suggestion though, if you don't like it, you can skip it.

If you do close the channel, you should probably check the `ok` return var when receiving on the channel in `makeTlfBreaksIfNeeded`.